Addie code challenge

addiebarron opened the pull request:
## Overview

- Implement the `parse` method
    `parserator_web/views.py` -> `AddressParse.parse`
- Complete the API endpoint
    `parserator_web/views.py` -> `AddressParse.get`
- Wire up the form to send requests to the API, display results from the API
    `parserator_web/templates/parserator_web/index.html` -> `extra_js`
- Add unit tests
    `tests/test_views.py` -> `test_api_parse_succeeds`, `test_api_parse_raises_error`

## Testing

* Run `docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app`

hancush commented on parserator_web/templates/parserator_web/index.html:
```diff
@@ -32,14 +31,111 @@ <h4>Parsing results</h4>
+      <div id="address-parse-error" class="result error">
+        <h4>Unable to parse that address: <span class="exception-name"></span></h4>
+      </div>
+      <div id="request-error" class="result error">
```
Love this! 

I have a small nitpick: If you run into a server error after you get a parse error, the parse error message sticks around, which is a little confusing.

![Screen Shot 2020-10-12 at 9 19 32 AM](https://user-images.githubusercontent.com/12176173/95757017-32032400-0c6c-11eb-8551-ce1252144954.png)

What do you think about updating the server error display logic to hide any other error messages?

addiebarron replied:
Great catch, I didn't check that case. I think my work in refactoring AJAX errors addressed this concern, but let me know if you can repro it again.

hancush commented on parserator_web/static/css/custom.css:
```diff
@@ -1 +1,13 @@
 /* Add custom style overrides here. */
+
+.result {
+    display: none;
```
+1 to factoring out inline styles! It's a little counterintuitive, I think, that the result class is hidden by default. Except for this initial state, you've hidden and shown results via jQuery. What do you think on consolidating result visibility in the JavaScript and setting initial visibility with `$.hide()`, instead of in the CSS?

addiebarron replied:
Definitely understand this concern. I set initial styles in the CSS in the hopes of avoiding a Flash of Unstyled Content, since some of those results elements have text initially. Did a quick test to show that that does indeed happen if `.hide()` is called on `document.ready`:

![fouc](https://user-images.githubusercontent.com/11164294/95932556-29a90700-0d9a-11eb-9cca-a1dac15fe0b1.gif)

I'm open to suggestions on getting all that logic into the JS without causing this issue!

hancush replied:
Tricky! Looks like total consolidation isn't a great option, then.

Bootstrap provides a `d-none` class but unfortunately it's marked as `!important`, so `hide()` / `show()` doesn't have the desired effect. 

A couple of ideas:

- Separating out `display: none` into its own class, and giving that class to initially hidden elements, could address my reservation about combining styles that won't always belong together under one class name.
- In the past, I've used `d-none`, and [adding and removing it](https://github.com/datamade/bga-payroll/blob/78aea94ef7c71db923d41a66684c323ee022682d/bga_database/static/js/employer.js#L137-L135) instead of using `hide()` / `show()`.

No required changes here, just wanted to drop a note in this thread. 🙃

hancush commented on parserator_web/views.py:
```diff
@@ -14,11 +14,38 @@ class AddressParse(APIView):
+            address_components, address_type = self.parse(address);
+
+            # Return the results of a successful parse
+            response = {
```
We don't (yet) have a preferred style for API responses, but one general rule of thumb is that it's friendlier to provide a consistent response format, whether the request succeeds or fails. That way, you can explicitly check whether the request succeeded, rather than inferring from the absence of an error key. 

How about revising your response dictionaries to contain a `status` key with a value of the appropriate HTTP status code? 

(Jean shared this article with me when I put the question of preferred style to the DataMade team: https://www.baeldung.com/rest-api-error-handling-best-practices. Maybe you'll find it helpful, too!)

addiebarron replied:
Appreciate this feedback and the resource! After hearing those thoughts, I thought a decent way to clean this up would be to send a standard HTTP status code with the response, handle all AJAX errors together, and distinguish between internal errors and other statuses once the AJAX error has been caught. Let me know what you think of the changes in 93f6821fd87d8a627211dcfd8f223669901c2860.

hancush commented on parserator_web/views.py:
```diff
@@ -14,11 +14,38 @@ class AddressParse(APIView):
+        # Grab the address from our request data
+        address = request.query_params.get('address');
+
+        try:
```
When writing `try / except` blocks, it's more explicit and less error prone to include only the lines that could raise exceptions you want to handle in the `try` block. For example, if you had a typo in your response dictionary, it would be handled the same as a parse error, which can lead to some baffling debugging. How about moving the creation of a successful response dictionary [to an `else` block](https://docs.python.org/3/tutorial/errors.html#handling-exceptions)?

addiebarron replied:
Thanks! I've been deep in JS world lately so I had completely forgotten about `else` blocks. Great solution to an issue I hadn't considered. Just implemented this change alongside the API response changes above.

hancush commented on tests/test_views.py:
```diff
@@ -1,15 +1,32 @@
-    pytest.fail()
+    
+    response = client.get('/api/parse/', {'address': address_string})
+    # Could this call ^ be fixturized for DRYer code? 
```
Technically, yes! You could use [parameterization](https://github.com/datamade/testing-guidelines/blob/master/intermediate-python-testing.md#parameterizing-fixtures) to pull the address string out into a variable. Here's [an example in the wild](https://github.com/datamade/la-metro-councilmatic/blob/23e1dccf8c60f7cf2c3f5624c05b28b233b7e25c/tests/test_events.py#L59-L93) of conditionally performing test assertions, depending on parameter values. 

I think whether you'd want to combine success and failure cases is a matter of preference and context. It isn't always the best separation of concerns, although [pytest's docs on params include an example with mixed outcomes](https://docs.pytest.org/en/stable/example/parametrize.html#parametrizing-conditional-raising), but it can be a good strategy for reducing duplicate code if, as in these tests, most of the logic is the same between test cases.

I'll leave it to you whether you want to take a stab at it!

addiebarron replied:
Thanks! Making a wild guess that the majority apps will need more than two test cases, so parametrizing seems like a good path forward in any case. I'm now questioning whether the entire response structure needs to be tested against, but I decided to keep that aspect of the tests in the interest of keeping the commit slim. Also opted to leave out fixtures to limit scope. Curious to hear your thoughts.

hancush reviewed (changes requested):
Hey, @addiebarron! I'm Hannah, and I'll be reviewing your code challenge and co-interviewing you with Derek this week. 

There's a lot of great stuff going on here. First and foremost, the app works great! At the code level, I love to see your use of `async` / `await` (we've been working as a company to transition to contemporary JavaScript syntax for new projects), and that you thought to handle AJAX errors. I left a couple of comments inline for your attention.

We'd like you to respond to the review and make the requested changes before your interview. For guidance, we like and try to stick to [Thoughtbot's guide to code review](https://github.com/thoughtbot/guides/tree/master/code-review) when working together on pull requests.

Let me know in the relevant thread if you have any questions, etc., and I'll be sure to get back to you ASAP. Look forward to talking more on Wednesday.

addiebarron commented:
Hey Hannah,

Thanks so much for the thoughtful and personable review! I did my best to address the issues you raised and incorporate the ideas you brought to the table. Added these changes in 93f6821fd87d8a627211dcfd8f223669901c2860:

`views.py` 
- Refactored API response generation to return a status code along with the response data
- Moved all code that isn't expected to generate an exception into an `else` block.

`index.html` 
- Refactored Javascript to handle all AJAX errors in one handler, which makes use of the platform's HTTP status codes and leans more on structure of jQuery AJAX calls.
- In the course of the above JS refactoring, fixed an issue that caused the last results to stick around when a server error is displayed.

`tests.py`
- Parametrized the API test functions for scalability and learning purposes :)

Very much looking forward to discussing tomorrow.
